### PR TITLE
Add pagination for GET /api/v1/jobs

### PR DESCRIPTION
**Pull request by asdil12**
https://progress.opensuse.org/issues/121102

### CODE REVIEW

**Review by kraih — DISMISSED** [review]
You forgot pagination tests with the `Link` headers, like [these](https://github.com/os-autoinst/openQA/blob/c73687c22fdee12430e1dab26365347517101840/t/api/11-bugs.t#L91-L139).

**Comment by asdil12** [comment]
> You forgot pagination tests with the `Link` headers, like [these](https://github.com/os-autoinst/openQA/blob/c73687c22fdee12430e1dab26365347517101840/t/api/11-bugs.t#L91-L139).

I added tests.
Please have a look at how I don't do the pagination properly for the `latest=1` case which would be hard due to the deduplication breaking the `has_more` check.
Maybe you have an idea how this could be handled.
I'm not sure if pagination in that case is actually needed, though.

**Comment by kraih** [comment]
> Please have a look at how I don't do the pagination properly for the `latest=1` case which would be hard due to the deduplication breaking the `has_more` check. Maybe you have an idea how this could be handled. I'm not sure if pagination in that case is actually needed, though.

If none of the global limits applied to it before either, then it's probably fine. Or it was an oversight. 🤷

**Comment by asdil12** [comment]
well if I understand it correctly, the global limit did apply before as well as the page, but the page could have less than `$limit` entries - in theory even 0, due to the subquery and the deduplication in perl code.

**Review by okurz — CHANGES_REQUESTED** [review]
The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

**Comment by asdil12** [comment]
> The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

Not sure what you mean - the commit message has a ticket reference and a descriptive message.
If you want something else, please elaborate.

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm [hunk 1]
```diff
@@ -82,18 +82,22 @@ sub list ($self) {
     $validation->optional('scope')->in('current', 'relevant');
     $validation->optional('limit')->num(0);
     $validation->optional('latest')->num(1);
+    $validation->optional('limit')->num;
```
Comment (kalikiana) [inline lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:85]: This is duplicated

**Comment by kalikiana** [comment]
> Not sure what you mean - the commit message has a ticket reference and a descriptive message. If you want something else, please elaborate.

The commit message **only** contains a URL. Generally it's good to have at least a sentence about what is being implemented.

If it was me I'd probably write something like this:

```
By handling `row` and `offset` parameters a subset of jobs can be queried. This relies on on the existing `pagination_links_header` feature.

See: https://progress.opensuse.org/issues/121102
```

**Comment by okurz** [comment]
> Not sure what you mean - the commit message has a ticket reference and a descriptive message. If you want something else, please elaborate.

Well, for example explain the reasoning and what design choices you made. Doesn't have to be anything fancy. Something like "As already done in commit ... adding pagination for /api/v1/jobs to prevent server overloading", etc.

File: t/api/04-jobs.t [hunk 2]
```diff
@@ -144,19 +144,78 @@ subtest 'argument combinations' => sub {
...
+        ok $links->{first}, 'has first page';
+        ok !$links->{next}, 'no next page';
+        ok $links->{prev}, 'has previous page';
+
+        $t->get_ok('/api/v1/jobs?limit=18&latest=1')->status_is(200);
+        my @jids_l = ();
+        my @subjids_l = ();
+        for my $j (@{$t->tx->res->json->{jobs}}) {
+            push(@jids_l, $j->{id});
+        }
```
Comment (okurz) [inline t/api/04-jobs.t:215]: duplicated

File: t/api/04-jobs.t [hunk 3]
```diff
@@ -144,19 +144,78 @@ subtest 'argument combinations' => sub {
...
+        ok $links->{next}, 'has next page';
+        ok $links->{prev}, 'has previous page';
+
+        @subjids = @jids[@jids - 18 .. $#jids - 15];
+        $t->get_ok($links->{next}{link})->status_is(200)->json_is("/jobs/0/id" => $subjids[0])
+          ->json_is("/jobs/1/id" => $subjids[1])->json_is("/jobs/2/id" => $subjids[2])->json_hasnt("/jobs/3");
+        $links = $t->tx->res->headers->links;
+        ok $links->{first}, 'has first page';
+        ok !$links->{next}, 'no next page';
+        ok $links->{prev}, 'has previous page';
```
Comment (okurz) [inline t/api/04-jobs.t:208]: this looks a bit duplicated and too much reliant on specific index position in arrays. That should be cleaned up a bit although functionally looks fine

**Review by okurz — CHANGES_REQUESTED** [review]

**Review by Martchus — COMMENTED** [review]
Generally ok but consider looking into the points mentioned by @okurz .

**Comment by kraih** [comment]
Ticket has been reassigned to me, guess i'll clean up the PR.

File: lib/OpenQA/Schema/ResultSet/Jobs.pm [hunk 4]
```diff
@@ -84,8 +84,11 @@ MACHINE. For each set of dupes, only the latest job found is included in
 the return array.
 
 =cut
-sub latest_jobs ($self, $until = undef) {
-    my @jobs = $self->search($until ? {t_created => {'<=' => $until}} : undef, {order_by => ['me.id DESC']});
+sub latest_jobs ($self, $until = undef, $limit = undef, $offset = undef) {
+    my %attrs = (order_by => ['me.id DESC']);
+    $attrs{rows} = $limit if defined $limit;
+    $attrs{offset} = $offset if defined $offset;
+    my @jobs = $self->search($until ? {t_created => {'<=' => $until}} : undef, \%attrs);
```
Comment (kraih) [inline lib/OpenQA/Schema/ResultSet/Jobs.pm:91]: This was actually dead code.

**Pull request closed without merging**